multicluster/controllers/multicluster/member/clusterset_controller.go

```diff
@@ -90,60 +95,110 @@ func NewMemberClusterSetReconciler(client client.Client,
 
-	// Handle create or update
-	if r.clusterSetConfig == nil {
-		r.clusterID, err = common.GetClusterID(r.ClusterCalimCRDAvailable, req, r.Client, clusterSet)
-		if err != nil {
-			return ctrl.Result{}, err
+		// Handle create or update
+
+		newLeader := clusterSet.Spec.Leaders[0]
+
+		clusterSetCreated = r.clusterID != common.ClusterID(clusterSet.Spec.ClusterID) || r.clusterSetID != common.ClusterSetID(clusterSet.Name)
+		leaderChanged := r.installedLeader.clusterID != newLeader.ClusterID || r.installedLeader.serverUrl != newLeader.Server ||
+			r.installedLeader.secretName != newLeader.Secret
+		r.clusterSetConfig = clusterSet.DeepCopy()
+
+		if !leaderChanged && !clusterSetCreated {
+			klog.V(2).InfoS("No change for leader cluster configuration")
+			return nil
 		}
-		r.clusterSetID = common.ClusterSetID(clusterSet.Name)
-		if clusterSet.Spec.ClusterID == "" {
-			// ClusterID is a required feild, and the empty value case should only happen
-			// when Antrea Multi-cluster is upgraded from an old version prior to v1.13.
-			// Here we try to update the ClusterSet's ClusterID when it's configured in an
-			// existing ClusterClaim.
-			clusterSet.Spec.ClusterID = string(r.clusterID)
-			err = r.Update(context.TODO(), clusterSet)
+
+		if clusterSetCreated {
+			// ClusterSet deletion may fail and retry, but a ClusterSet may have been created just right before the retry.
+			// In that case, ClusterSet deletion retry will be like an update action, so try to delete stale resources
+			// here before initilizing a new ClusterSet.
+			if err := r.cleanUpResources(ctx); err != nil {
+				return err
+			}
+
+			r.clusterID, err = common.GetClusterID(r.ClusterCalimCRDAvailable, req, r.Client, clusterSet)
 			if err != nil {
-				klog.ErrorS(err, "Failed to update ClusterSet's ClusterID", "clusterset", req.NamespacedName)
-				return ctrl.Result{}, err
+				return err
+			}
+			r.clusterSetID = common.ClusterSetID(clusterSet.Name)
+			if clusterSet.Spec.ClusterID == "" {
+				// ClusterID is a required feild, and the empty value case should only happen
+				// when Antrea Multi-cluster is upgraded from an old version prior to v1.13.
+				// Here we try to update the ClusterSet's ClusterID when it's configured in an
+				// existing ClusterClaim.
+				clusterSet.Spec.ClusterID = string(r.clusterID)
+				err = r.Update(context.TODO(), clusterSet)
+				if err != nil {
+					klog.ErrorS(err, "Failed to update ClusterSet's ClusterID", "clusterset", req.NamespacedName)
+					return err
+				}
 			}
 		}
+		return r.createRemoteCommonArea(clusterSet)
 	}
-	r.clusterSetConfig = clusterSet.DeepCopy()
 
-	err = r.createOrUpdateRemoteCommonArea(clusterSet)
-	if err != nil {
+	if err := processClusterSet(); err != nil {
 		return ctrl.Result{}, err
 	}
 
+	if clusterSetCreated {
+		// The CommonArea creation succeeded here and so notify StaleController to
+		// clean up stale imported resources and ResourceExports.
+		r.commonAreaCreationCh <- struct{}{}
```

Comment:
Note `r.commonAreaCreationCh` is an unbuffered channel, do we want to block here if staleController is stuck in doing some cleanup?
If it's just to make sure staleController will perform cleanup after this notication, perhaps it should have one buffer, and just ignore sending the notification when the buffer is full.

Reply:
Done.